File: app/src/main/java/com/example/initialproject/model/RequestModelImpl.java
package com.example.initialproject.model;

import android.support.annotation.LongDef;
import android.support.annotation.NonNull;
import android.util.Log;

import java.io.IOException;

import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Retrofit;

public class RequestModelImpl implements RequestModel {

    @Override
    public void getData(final RequestListener requestListener, final String url, int type) {
        Retrofit retrofit = new Retrofit.Builder().baseUrl("https://news-at.zhihu.com/api/3/news/").build();
        Api api = retrofit.create(Api.class);
        Call<ResponseBody> call = null;
        if (type == 1) call = api.getLatestData(url);
        else call = api.getAddData(url);
        Log.d("Impor", "访问" + url);
        call.enqueue(new retrofit2.Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, retrofit2.Response<ResponseBody> response) {
                if (response.isSuccessful()) {
                    try {
                        requestListener.onSuccess(response.body().string());
                    } catch (IOException e) {
                        e.printStackTrace();
                        requestListener.onFail();
                    }
                } else {
                    requestListener.onFail();
                }
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                requestListener.onFail();
            }
        });
    }
}

File: app/src/main/java/com/example/initialproject/model/RequestModelImpl.kt
package com.example.initialproject.model

import kotlinx.coroutines.experimental.launch
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Retrofit
import java.io.IOException

open class RequestModelImpl : RequestModel {
    override fun getData(requestListener: com.example.initialproject.model.RequestListener, url: String, type: Int) {
        val api = Retrofit
                .Builder()
                .baseUrl("https://news-at.zhihu.com/api/3/news/")
                .build()
                .create(Api::class.java)
        val call: Call<ResponseBody>?
        call = if (type == 1)
            api.getLatestData(url)
        else
            api.getAddData(url)
        launch {//协程
            val response: ResponseBody?
            try {
                response = call.execute().body()
                requestListener.onSuccess(response.string())
            } catch (e: IOException) {
                e.printStackTrace()
                requestListener.onFail()
            }
        }
    }
}